packages/vitest/src/node/plugins/globalSetup.ts

```diff
@@ -8,12 +10,23 @@ interface GlobalSetupFile {
 }
 
 async function loadGlobalSetupFiles(server: ViteDevServer): Promise<GlobalSetupFile[]> {
+  const node = new ViteNodeServer(server)
```

Comment:
can't you reuse `vitenode` from `ctx`? 

- https://github.com/vitest-dev/vitest/blob/main/packages/vitest/src/node/plugins/index.ts#L9
- https://github.com/vitest-dev/vitest/blob/main/packages/vitest/src/node/core.ts#L84

I think it's undefined at the initialization, but you can pass the whole context to `GlobalPlugin`, I think

Reply:
loadGlobalSetupFiles is called in buildStart hook, so undefined at init wouldn't be an issue.
https://github.com/vitest-dev/vitest/blob/c9f8aad7d1eb9705f99675966b428497ae518505/packages/vitest/src/node/plugins/globalSetup.ts#L67

i'm not sure how to get/pass ctx into GlobalPlugin though, can you elaborate on that? 

got the code to init the ViteNodeRunner from here https://github.com/vitest-dev/vitest/blob/568668ae747e3cbe162b7b13dc6e259655c6c313/packages/vite-node/src/cli.ts#L77

Reply:
Is that possible to run the global setup after the vite server starts? Or what will that cause?

Reply:
globalSetup should run once before any testfile is executed and teardown once after all testfiles are done. As long as that is true, it's ok to not execute it on vite server start or even before (that would make it impossible to use the vite server for loading the files anyways).

Reply:
> loadGlobalSetupFiles is called in buildStart hook, so undefined at init wouldn't be an issue.
> 
> https://github.com/vitest-dev/vitest/blob/c9f8aad7d1eb9705f99675966b428497ae518505/packages/vitest/src/node/plugins/globalSetup.ts#L67
> 
> i'm not sure how to get/pass ctx into GlobalPlugin though, can you elaborate on that?
> 
> got the code to init the ViteNodeRunner from here
> 
> https://github.com/vitest-dev/vitest/blob/568668ae747e3cbe162b7b13dc6e259655c6c313/packages/vite-node/src/cli.ts#L77

You can pass ctx to your plugin when initializing VitePlugins:

https://github.com/vitest-dev/vitest/blob/568668ae747e3cbe162b7b13dc6e259655c6c313/packages/vitest/src/node/plugins/index.ts#L9
https://github.com/vitest-dev/vitest/blob/568668ae747e3cbe162b7b13dc6e259655c6c313/packages/vitest/src/node/plugins/index.ts#L58

Reply:
ahh, yes that works although it only saves a little bit of code because the runner would still have to be inizialized. Maybe it is even preferable to use a separate instance to avoid globalSetup polluting the main instance?

On the flipside maybe we could/should pass the ctx down into setup so that it can be used in there and also maybe as a possible backchannel to add something to test contexts?

Reply:
> ahh, yes that works although it only saves a little bit of code because the runner would still have to be inizialized. Maybe it is even preferable to use a separate instance to avoid globalSetup polluting the main instance?
> 
> On the flipside maybe we could/should pass the ctx down into setup so that it can be used in there and also maybe as a possible backchannel to add something to test contexts?

Well, it at least should use test config, I think. Or otherwise `deps` and `transformMode` will have no effect in user config.

https://github.com/vitest-dev/vitest/blob/de3328485fdfaa76f4f6a32b6dd7d2b3bb954f84/packages/vitest/src/node/core.ts#L84